Source/WebCore/rendering/style/StyleCustomPropertyData.h

```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
-    }
+    unsigned size() const { return m_sizeExcludingInheritedValues + (m_inheritedValues ? m_inheritedValues->m_sizeExcludingInheritedValues : 0); }
 
-    CustomPropertyValueMap values;
+    void forEach(const Function<void(const KeyValuePair<AtomString, RefPtr<const CSSCustomPropertyValue>>&)>&) const;
+    AtomString keyAtIndex(unsigned) const;
 
 private:
     StyleCustomPropertyData() = default;
-    StyleCustomPropertyData(const StyleCustomPropertyData& other)
-        : RefCounted<StyleCustomPropertyData>()
-        , values(other.values)
-    { }
+    StyleCustomPropertyData(const StyleCustomPropertyData&);
+
+    RefPtr<const StyleCustomPropertyData> m_inheritedValues;
+    CustomPropertyValueMap m_values;
+    unsigned m_sizeExcludingInheritedValues;
 };
 
+inline StyleCustomPropertyData::StyleCustomPropertyData(const StyleCustomPropertyData& other)
+    : RefCounted<StyleCustomPropertyData>()
+    , m_inheritedValues(!other.m_inheritedValues && !other.m_values.isEmpty() ? &other : other.m_inheritedValues.get())
+    , m_values(!other.m_inheritedValues && !other.m_values.isEmpty() ? CustomPropertyValueMap() : other.m_values)
+{
+}
+
+inline const CSSCustomPropertyValue* StyleCustomPropertyData::get(const AtomString& name) const
+{
+    if (auto* value = m_values.get(name))
+        return value;
+    if (m_inheritedValues) {
+        ASSERT(!m_inheritedValues->m_inheritedValues);
+        if (auto* value = m_inheritedValues->m_values.get(name))
+            return value;
+    }
+    return nullptr;
+}
+
+inline void StyleCustomPropertyData::set(const AtomString& name, Ref<const CSSCustomPropertyValue>&& value)
+{
+    auto addResult = m_values.set(name, WTFMove(value));
+    if (addResult.isNewEntry && (!m_inheritedValues || !m_inheritedValues->m_values.contains(name)))
+        ++m_sizeExcludingInheritedValues;
+}
+
+inline bool StyleCustomPropertyData::operator==(const StyleCustomPropertyData& other) const
+{
+    if (size() != other.size())
+        return false;
+
+    for (auto& entry : m_values) {
+        auto* otherValue = other.get(entry.key);
+        if (!otherValue || !entry.value->equals(*otherValue))
+            return false;
+    }
+
+    if (m_inheritedValues && m_inheritedValues != other.m_inheritedValues) {
+        ASSERT(!m_inheritedValues->m_inheritedValues);
+        for (auto& entry : m_inheritedValues->m_values) {
+            if (m_values.contains(entry.key))
+                continue;
+            auto* otherValue = other.get(entry.key);
+            if (!otherValue || !entry.value->equals(*otherValue))
+                return false;
+        }
+    }
+
+    return true;
+}
+
+inline void StyleCustomPropertyData::forEach(const Function<void(const KeyValuePair<AtomString, RefPtr<const CSSCustomPropertyValue>>&)>& callback) const
+{
+    if (m_inheritedValues) {
+        ASSERT(!m_inheritedValues->m_inheritedValues);
+        for (const auto& iter : m_inheritedValues->m_values) {
```

Comment:
Can this have another name? We frown on abbreviations and I don’t think `iter` is particularly good either. How about `inheritedValue`?

Reply:
Or `entry`.

Reply:
Also no need for the `const`.